vlog4j-core/src/test/java/org/semanticweb/vlog4j/core/reasoner/implementation/QueryAnswerSizeTest.java

```diff
@@ -0,0 +1,293 @@
+	private static final PositiveLiteral Rey = Expressions.makePositiveLiteral(predR, e, y);
+	private static final PositiveLiteral Rxd = Expressions.makePositiveLiteral(predR, x, d);
+	private static final PositiveLiteral Rxe = Expressions.makePositiveLiteral(predR, x, e);
+
+	@Test
+	public void noFactsnoRules() throws IOException {
+		final KnowledgeBase kb = new KnowledgeBase();
+		try (VLogReasoner reasoner = new VLogReasoner(kb)) {
+			reasoner.reason();
+			assertEquals(0, reasoner.queryAnswerSize(Px, true).getSize());
+			assertEquals(0, reasoner.queryAnswerSize(Qx, true).getSize());
+			assertEquals(0, reasoner.queryAnswerSize(Rxy, true).getSize());
+		}
+	}
+
+	@Test
+	public void noFactsUniversalRule() throws IOException {
+		final KnowledgeBase kb = new KnowledgeBase();
+		kb.addStatement(QxPx);
+		try (VLogReasoner reasoner = new VLogReasoner(kb)) {
+			reasoner.reason();
+			assertEquals(0, reasoner.queryAnswerSize(Px, true).getSize());
+			assertEquals(0, reasoner.queryAnswerSize(Qx, true).getSize());
+			assertEquals(0, reasoner.queryAnswerSize(Rxy, true).getSize());
+		}
+	}
+
+	@Test
+	public void noFactsExistentialRule() throws IOException {
+		final KnowledgeBase kb = new KnowledgeBase();
+		kb.addStatement(RxyQyPx);
+		try (VLogReasoner reasoner = new VLogReasoner(kb)) {
+			reasoner.reason();
+			assertEquals(0, reasoner.queryAnswerSize(Px, true).getSize());
+			assertEquals(0, reasoner.queryAnswerSize(Qx, true).getSize());
+			assertEquals(0, reasoner.queryAnswerSize(Rxy, true).getSize());
+		}
+	}
+
+	@Test
+	public void pFactsNoRules() throws IOException {
+		final KnowledgeBase kb = new KnowledgeBase();
+		kb.addStatements(factPc, factPd);
+		try (VLogReasoner reasoner = new VLogReasoner(kb)) {
+			reasoner.reason();
+			assertEquals(2, reasoner.queryAnswerSize(Px, true).getSize());
+			assertEquals(0, reasoner.queryAnswerSize(Qx, true).getSize());
+			assertEquals(0, reasoner.queryAnswerSize(Rxy, true).getSize());
+		}
+	}
+
+	@Test
+	public void pFactsUniversalRule() throws IOException {
+		final KnowledgeBase kb = new KnowledgeBase();
+		kb.addStatements(factPc, factPd, QxPx);
+		try (VLogReasoner reasoner = new VLogReasoner(kb)) {
+			reasoner.reason();
+			assertEquals(2, reasoner.queryAnswerSize(Px, true).getSize());
+			assertEquals(2, reasoner.queryAnswerSize(Px, false).getSize());
+			assertEquals(2, reasoner.queryAnswerSize(Qx, true).getSize());
+			assertEquals(2, reasoner.queryAnswerSize(Qx, false).getSize());
+			assertEquals(0, reasoner.queryAnswerSize(Rxy, true).getSize());
+			assertEquals(0, reasoner.queryAnswerSize(Rxy, false).getSize());
+			assertEquals(1, reasoner.queryAnswerSize(factPc, false).getSize());
+			assertEquals(1, reasoner.queryAnswerSize(factPd, false).getSize());
+			assertEquals(1, reasoner.queryAnswerSize(factQc, false).getSize());
+			assertEquals(1, reasoner.queryAnswerSize(factQd, false).getSize());
+		}
+	}
+
```

Comment:
I would also add tests for a query with no variables, for which
- the predicate is not in the KB
- the predicate is in the KB, but there is no query answer
(this is more of a test for vLog then a unit test of our method, but it doesn't hurt)

Reply:
The first three test cover these cases

Reply:
but in these first three tests, the queries have variables. I was referring to queries that only contain constants.